### ISSUE TITLE
Style/ClassEqualityComparison ignores AllowedMethods

### ISSUE THREAD

**Issue by akunszt** [rubocop/rubocop#11914]
## Expected behavior

Rubocop should not ignore settings in `AllowedMethods` for `Style/ClassEqualityComparison`. I think I configured it correctly in .rubocop.yml, but if I made a mistake then sorry for the noise.

## Actual behavior

I set `AllowedMethods` to include `==` but it looks like RuboCop ignores that settings and keep complaining about the same offense I want to ignore.
```
/tmp/rubocop $ rubocop --only Style/ClassEqualityComparison --debug -f simple
For /tmp/rubocop: configuration from /tmp/rubocop/.rubocop.yml
Default configuration from /usr/lib64/ruby/gems/3.1.0/gems/rubocop-1.52.0/config/default.yml
Scanning /tmp/rubocop/test.rb
Loading cache from /home/arpad/.cache/rubocop_cache/b446cca981d8019a5b960236f5d9f52857607a4c/70e37c62e63f44962d6ae11027b6808747bb3bff/a89a5c8ae04809310731d497861eb10987076240
== test.rb ==
C:  6: 31: Style/ClassEqualityComparison: Use instance_of? instead of comparing classes.

1 file inspected, 1 offense detected
```

## Steps to reproduce the problem

Create a simple `.rubocop.yml` (enabling or disabling NewCops doesn't matter, I left it there to suppress unrelated warnings):
```yaml
---
AllCops:
  NewCops: enable

Style/ClassEqualityComparison:
  AllowedMethods:
    - "=="
```

And a short test code, name it `test.rb`.
```ruby
class_name = ARGV.first
s = 'Hello world'

puts "Yes it's a string" if s.class.to_s == class_name
```

It is a valid Ruby code (doesn't have a lot of meaning but valid), and you can execute it safely with existing or non-existing class names.
```
/tmp/rubocop $ ruby test.rb Integer
/tmp/rubocop $ ruby test.rb String
Yes it's a string
/tmp/rubocop $ ruby test.rb FooBar
```

Execute the following command to run RuboCop.
```
$ rubocop --only Style/ClassEqualityComparison --debug -f simple
```

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
$ rubocop -V
1.52.0 (using Parser 3.2.2.1, rubocop-ast 1.29.0, running on ruby 3.1.4) [x86_64-linux]
```

**Comment by akunszt** [comment]
I did a very short and small investigation. I found that here the `def_node` is nil, so the `allowed_method?(...)` never reached:
https://github.com/rubocop/rubocop/blob/e216c9baf40e37dbe701524f214d36eb8f264010/lib/rubocop/cop/style/class_equality_comparison.rb#L87-L92

The "contents" in `node.each_ancestor()` looks similar to the test code I provided but I never touched RuboCops internals so I am not sure what am I seeing there (tokenizer output?). But the `node.each_ancestor( :def, :defs )` is an empty.

```
[s(:if,
  s(:send,
    s(:send,
      s(:send,
        s(:lvar, :s), :class), :to_s), :==,
    s(:lvar, :class_name)),
  s(:send, nil, :puts,
    s(:str, "Yes it's a string")), nil),
 s(:begin,
  s(:lvasgn, :class_name,
    s(:send,
      s(:const, nil, :ARGV), :first)),
  s(:lvasgn, :s,
    s(:str, "Hello world")),
  s(:if,
    s(:send,
      s(:send,
        s(:send,
          s(:lvar, :s), :class), :to_s), :==,
      s(:lvar, :class_name)),
    s(:send, nil, :puts,
      s(:str, "Yes it's a string")), nil))]
```

**Comment by koic** [comment]
The cop's `AllowedMethods` examples are mis-documented for how to use it. I opened #11917 to fix it.

**Comment by akunszt** [comment]
Thanks for the answer. If I understand this correctly then with this cop you cannot allow to match an object class to a String. So if I receive that String during run-time then my only choice is to disable this cop, is it correct?

**Comment by koic** [comment]
I think that case can be solved by using `Module.const_get`:

```diff
 class_name = ARGV.first
 s = 'Hello world'

-puts "Yes it's a string" if s.class.to_s == class_name
+puts "Yes it's a string" if s.instance_of?(Module.const_get(class_name))
```

In this case, RuboCop's static analysis can't recognize the variable type, so it doesn't support autocorrection, so you'll have to correct it by handcraft.

**Comment by akunszt** [comment]
Those are not the exactly same.

```
$ ruby test.rb String
Yes it's a string
$ ruby test.rb FooBar
$ ruby test_with_module_const.rb String
Yes it's a string
$ ruby test_with_module_const.rb FooBar
test_with_module_const.rb:4:in `<main>': uninitialized constant Module::FooBar (NameError)

puts "Yes it's a string" if s.instance_of?( Module.const_get( class_name ) )
 
```
My real code can receive strings which doesn't contain a real class name. Yes, I can put it into a `begin ... rescue ... end` in this case(how fast is that BTW?) but for the sake readability it is easier just to disable this cop.

**Comment by koic** [comment]
Yeah, If you don't like it, you can use `rubocop:disable` and etc.
